### PR TITLE
[FIX] CsiFingerIdMzTAbWriter, SiriusMzTabWriter 

### PR DESCRIPTION
**Pull request by oliveralka**
Feature_id was not reported correctly in the CsiFingerIdMzTabWriter.

Use the number of candidates for Sirius Output.
Use the number of top hits for CSIFingerID Output.

Fix corresponding CSIFingerID-test.

### CODE REVIEW

**Comment by oliveralka** [comment]
Failed test:
`projectroot.src.tests.class_tests.openms.File_test`
seems unrelated

**Comment by oliveralka** [comment]
rebuild jenkins

File: src/utils/SiriusAdapter.cpp [hunk 1]
```diff
@@ -199,7 +199,7 @@ class TOPPSiriusAdapter :
     bool no_mt_info = getFlag_("no_masstrace_info_isotope_pattern");
 
     // needed for counting
-    int top_n_hits = getIntOption_("top_n_hits"); 
+    int top_n_hits = getIntOption_("top_n_hits") + 1;
```
Comment (timosachsenberg) [inline src/utils/SiriusAdapter.cpp]: why +1?

File: src/utils/SiriusAdapter.cpp [hunk 2]
```diff
@@ -209,7 +209,7 @@ class TOPPSiriusAdapter :
     const QString isotope = getStringOption_("isotope").toQString();
     const QString noise = QString::number(getIntOption_("noise"));
     const QString ppm_max = QString::number(getIntOption_("ppm_max"));
-    const QString candidates = QString::number(getIntOption_("candidates"));
+    const QString candidates = QString::number(getIntOption_("candidates")+1);
```
Comment (timosachsenberg) [inline src/utils/SiriusAdapter.cpp]: why +1

Comment (timosachsenberg) [inline src/utils/SiriusAdapter.cpp]: better no to use a string here but a Size type.
Conversion to string can be done if required in line 311

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp [hunk 3]
```diff
@@ -68,7 +67,8 @@ void CsiFingerIdMzTabWriter::read(const std::vector<String> & sirius_output_path
     
       if (rowcount > 1)
       {
-        
+        const UInt top_n_hits_cor = (top_n_hits >= rowcount) ? rowcount : top_n_hits;
```
Comment (lukaszimmermann) [inline src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: Maybe use `std::min`. Comment what happens here. Variable name unclear. (cor?)

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp [hunk 4]
```diff
@@ -89,7 +89,6 @@ void CsiFingerIdMzTabWriter::read(const std::vector<String> & sirius_output_path
         // results from scan were not assigned to a feautre
         if (feature_id == "id_0") {feature_id = "null";}
```
Comment (lukaszimmermann) [inline src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: would not change the value of `feature_id`. Perhaps only use the "null" value where `feature_id` is read, otherwise I'd use a new variable name.

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp [hunk 5]
```diff
@@ -89,7 +89,6 @@ void CsiFingerIdMzTabWriter::read(const std::vector<String> & sirius_output_path
...
 
-        const UInt top_n_hits_cor = (top_n_hits > rowcount) ? rowcount : top_n_hits;
         for (Size j = 1; j < top_n_hits_cor; ++j)
```
Comment (lukaszimmermann) [inline src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: What about `j = 0`?

File: src/utils/SiriusAdapter.cpp [hunk 6]
```diff
@@ -151,7 +151,7 @@ class TOPPSiriusAdapter :
     // internal sirius parameters
     registerStringOption_("profile", "<choice>", "qtof", "Specify the used analysis profile", false);
     setValidStrings_("profile", ListUtils::create<String>("qtof,orbitrap,fticr"));
-    registerIntOption_("candidates", "<num>", 5, "The number of candidates in the output.", false);
+    registerIntOption_("candidates", "<num>", 5, "The number of candidates in the sirius output.", false);
```
Comment (lukaszimmermann) [inline src/utils/SiriusAdapter.cpp]: sirius -> SIRIUS

File: src/utils/SiriusAdapter.cpp [hunk 7]
```diff
@@ -161,7 +161,7 @@ class TOPPSiriusAdapter :
     registerStringOption_("elements", "<choice>", "CHNOP[5]S[8]Cl[1]", "The allowed elements. Write CHNOPSCl to allow the elements C, H, N, O, P, S and Cl. Add numbers in brackets to restrict the maximal allowed occurrence of these elements: CHNOP[5]S[8]Cl[1].", false);
     registerIntOption_("compound_timeout", "<num>", 10, "Time out in seconds per compound. To disable the timeout set the value to 0", false);
     registerIntOption_("tree_timeout", "<num>", 0, "Time out in seconds per fragmentation tree computation.", false);
-    registerIntOption_("top_n_hits", "<num>", 10, "The top_n_hit for each compound written to the output", false);
+    registerIntOption_("top_n_hits", "<num>", 10, "The top_n_hit for each compound written to the CSI:FingerID output", false);
```
Comment (lukaszimmermann) [inline src/utils/SiriusAdapter.cpp]: Maybe: The number of top hits for each compound written...

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp (hunk 4 above)
Comment (oliveralka) [reply src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: Right now, feature_id is processed from name of the .csv file.
If no feature_id was associated the regex will still pick up an id_0. 
The writer will then write a "null" in the feautre_id column. 
In the case a feature_id is present it should look like:
id_3456783223456734

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp (hunk 5 above)
Comment (oliveralka) [reply src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: +1

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp [hunk 8]
```diff
@@ -64,10 +64,13 @@ void CsiFingerIdMzTabWriter::read(const std::vector<String> & sirius_output_path
     {
       CsvFile compounds(pathtocsicsv, '\t');
       const UInt rowcount = compounds.rowCount();
-    
+
       if (rowcount > 1)
       {
-        const UInt top_n_hits_cor = (top_n_hits >= rowcount) ? rowcount : top_n_hits;
+        // correction if the rowcount is smaller than the number of hits used as parameter
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: Minor suggestion as I found the comments/variable names could be a bit more descriptive.
you could define
``` 
const UInt n_header_rows = 1;
const UInt n_hits_rows = rowcount - n_header_rows;
and
const UInt n_hits_to_extract = std::min(top_n_hits, n_hits_rows);
```
instead of using the top_n_hits_cor variable name .
comment in line 92 would probably than not be necessary
and you could loop from j=0 to n_hits_to_extract but inside of the loop call .getRow(n_header_rows + j) instead.
All in all slightly clearer (at least to me) :P

File: src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp (hunk 4 above)
Comment (timosachsenberg) [reply src/openms/source/FORMAT/DATAACCESS/CsiFingerIdMzTabWriter.cpp]: I think Lukas means that it is confusing/hard to follow if the feature_id is overwritten.
I think what he means is to better set the id to e.g. "UNKNOWN_ID" (ideally a constant defined as static constexpr) instead of id_0.
When writing to mzTab you would check if it is  == this value and if so write out "null".

**Comment by timosachsenberg** [comment]
I will merge this PR as it contains a fix to a bug. As Oliver is on holidays, I would like to ask him to resolve the remaining minor comments in a new PR.
